File: Wire-iOS/Sources/UserInterface/ConversationList/ListContent/ConversationListCell.m
// 
// Wire
// Copyright (C) 2016 Wire Swiss GmbH
// 
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
// 
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU General Public License for more details.
// 
// You should have received a copy of the GNU General Public License
// along with this program. If not, see http://www.gnu.org/licenses/.
// 

#import "ConversationListCell.h"
#import "ConversationListCell+Internal.h"

#import "ConversationListItemView.h"



#import "Settings.h"

#import "MediaPlayer.h"


#import "AnimatedListMenuView.h"
#import "Wire-Swift.h"


@interface ConversationListCell () <AVSMediaManagerClientObserver>

@property (nonatomic) ConversationListItemView *itemView;

@property (nonatomic) NSLayoutConstraint *titleBottomMarginConstraint;

@property (nonatomic) id typingObserverToken;
@end

@interface ConversationListCell (Typing) <ZMTypingChangeObserver>
@end

@implementation ConversationListCell

- (void)dealloc
{
    [AVSMediaManagerClientChangeNotification removeObserver:self];
}

- (instancetype)initWithFrame:(CGRect)frame
{
    self = [super initWithFrame:frame];
    if (self) {
        [self setupConversationListCell];
    }
    return self;
}

- (void)setupConversationListCell
{
    self.separatorLineViewDisabled = YES;
    self.maxVisualDrawerOffset = MaxVisualDrawerOffsetRevealDistance;
    self.overscrollFraction = CGFLOAT_MAX; // Never overscroll
    self.canOpenDrawer = NO;
    self.clipsToBounds = YES;

    self.itemView = [[ConversationListItemView alloc] init];
    
    UITapGestureRecognizer *tapGestureRecognizer = [[UITapGestureRecognizer alloc] initWithTarget:self
                                                                                           action:@selector(onRightAccessorySelected:)];
    [self.itemView.rightAccessory addGestureRecognizer:tapGestureRecognizer];
    [self.swipeView addSubview:self.itemView];

    self.menuDotsView = [[AnimatedListMenuView alloc] init];
    [self.menuView addSubview:self.menuDotsView];
    
    [self setNeedsUpdateConstraints];
     
    [AVSMediaManagerClientChangeNotification addObserver:self];
}

- (void)setVisualDrawerOffset:(CGFloat)visualDrawerOffset updateUI:(BOOL)doUpdate
{
    [super setVisualDrawerOffset:visualDrawerOffset updateUI:doUpdate];
    
    // After X % of reveal we consider animation should be finished
    const CGFloat progress = (visualDrawerOffset / MaxVisualDrawerOffsetRevealDistance);
    [self.menuDotsView setProgress:progress animated:YES];
    if (progress >= 1 && ! self.overscrollStartDate) {
        self.overscrollStartDate = [NSDate date];
    }
    
    self.itemView.visualDrawerOffset = visualDrawerOffset;
}

- (void)setConversation:(ZMConversation *)conversation
{
    if (_conversation != conversation) {
        self.typingObserverToken = nil;
        _conversation = conversation;
        self.typingObserverToken = [_conversation addTypingObserver:self];
        
        [self updateAppearance];

        [self setupConversationObserverWithConversation: conversation];
    }
}
    
- (void)updateAppearance
{
    [self.itemView updateForConversation:self.conversation];
}
    
- (BOOL)canOpenDrawer
{
    return YES;
}

static CGSize cachedSize = {0, 0};

- (CGSize)sizeInCollectionViewSize:(CGSize)collectionViewSize
{
    if (!CGSizeEqualToSize(cachedSize, CGSizeZero) && cachedSize.width == collectionViewSize.width) {
        return cachedSize;
    }
        
    NSString *fullHeightString = @"Ü";
    [self.itemView configureWith:[[NSAttributedString alloc] initWithString:fullHeightString]
                        subtitle:[[NSAttributedString alloc] initWithString:fullHeightString attributes:[ZMConversation statusRegularStyle]]];
    
    CGSize fittingSize = CGSizeMake(collectionViewSize.width, 0);
    
    self.itemView.frame = CGRectMake(0, 0, fittingSize.width, 0);

    CGSize cellSize = [self.itemView systemLayoutSizeFittingSize:fittingSize];
    cellSize.width = collectionViewSize.width;
    cachedSize = cellSize;
    return cellSize;
}

+ (void)invalidateCachedCellSize
{
    cachedSize = CGSizeZero;
}

#pragma mark - AVSMediaManagerClientChangeNotification

- (void)mediaManagerDidChange:(AVSMediaManagerClientChangeNotification *)notification
{
    // AUDIO-548 AVMediaManager notifications arrive on a background thread.
    dispatch_async(dispatch_get_main_queue(), ^{
        if (notification.microphoneMuteChanged) {
            [self updateAppearance];
        }
    });
}

#pragma mark - DrawerOverrides

- (void)drawerScrollingStarts
{
    self.overscrollStartDate = nil;
}

@end


@implementation ConversationListCell (Typing)

- (void)typingDidChangeWithConversation:(ZMConversation *)conversation typingUsers:(NSSet<ZMUser *> *)typingUsers
{
    [self updateAppearance];
}

@end

File: wire-ios/Wire-iOS/Sources/UserInterface/ConversationList/ListContent/ConversationListCell.swift
//
// Wire
// Copyright (C) 2024 Wire Swiss GmbH
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program. If not, see http://www.gnu.org/licenses/.
//

import avs
import Foundation
import WireDesign
import WireSyncEngine

typealias MatcherConversation = Conversation & ConversationStatusProvider & TypingStatusProvider & VoiceChannelProvider

typealias ConversationListCellConversation = MatcherConversation & StableRandomParticipantsProvider

// MARK: - ConversationListCell

final class ConversationListCell: SwipeMenuCollectionCell,
    SectionListCellType {
    static let IgnoreOverscrollTimeInterval: TimeInterval = 0.005
    static let OverscrollRatio: CGFloat = 2.5

    static var cachedSize: CGSize = .zero

    var conversation: ConversationListCellConversation? {
        didSet {
            guard !(conversation === oldValue) else { return }

            typingObserverToken = nil
            if let conversation = conversation as? ZMConversation {
                typingObserverToken = conversation.addTypingObserver(self)
                setupConversationObserver(conversation: conversation)
            }

            updateAppearance()
        }
    }

    let itemView = ConversationListItemView()

    weak var delegate: ConversationListCellDelegate?

    private var typingObserverToken: Any?

    // MARK: - SectionListCellType

    var sectionName: String?
    var obfuscatedSectionName: String?
    var cellIdentifier: String?

    private var hasCreatedInitialConstraints = false
    let menuDotsView = AnimatedListMenuView()
    private var overscrollStartDate: Date?
    private var conversationObserverToken: Any?

    convenience init() {
        self.init(frame: .zero)
    }

    override init(frame: CGRect) {
        super.init(frame: frame)
        setupConversationListCell()
    }

    deinit {
        AVSMediaManagerClientChangeNotification.remove(self)
    }

    @available(*, unavailable)
    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }

    private func setupConversationListCell() {
        separatorLineViewDisabled = true
        maxVisualDrawerOffset = SwipeMenuCollectionCell.MaxVisualDrawerOffsetRevealDistance
        overscrollFraction = CGFloat.greatestFiniteMagnitude // Never overscroll
        clipsToBounds = true

        let tapGestureRecognizer = UITapGestureRecognizer(target: self, action: #selector(onRightAccessorySelected(_:)))
        itemView.rightAccessory.addGestureRecognizer(tapGestureRecognizer)
        swipeView.addSubview(itemView)

        menuView.addSubview(menuDotsView)

        setNeedsUpdateConstraints()

        AVSMediaManagerClientChangeNotification.add(self)
        backgroundColor = SemanticColors.View.backgroundConversationListTableViewCell
        addBorder(for: .bottom)
    }

    override func drawerScrollingEnded(withOffset offset: CGFloat) {
        if menuDotsView.progress >= 1 {
            var overscrolled = false
            if offset > frame.width / ConversationListCell.OverscrollRatio {
                overscrolled = true
            } else if let overscrollStartDate {
                let diff = Date().timeIntervalSince(overscrollStartDate)
                overscrolled = diff > ConversationListCell.IgnoreOverscrollTimeInterval
            }

            if overscrolled {
                delegate?.conversationListCellOverscrolled(self)
            }
        }
        overscrollStartDate = nil
    }

    override var accessibilityValue: String? {
        get {
            delegate?.indexPath(for: self)?.description
        }

        set {
            // no op
        }
    }

    override var accessibilityIdentifier: String? {
        get {
            identifier
        }

        set {
            // no op
        }
    }

    override var isSelected: Bool {
        didSet {
            if isIPadRegular() {
                itemView.selected = isSelected || isHighlighted
            }
        }
    }

    override var isHighlighted: Bool {
        didSet {
            if isIPadRegular() {
                itemView.selected = isSelected || isHighlighted
            } else {
                itemView.selected = isHighlighted
            }
        }
    }

    override func updateConstraints() {
        super.updateConstraints()

        if hasCreatedInitialConstraints {
            return
        }
        hasCreatedInitialConstraints = true

        [itemView, menuDotsView, menuView].forEach { $0.translatesAutoresizingMaskIntoConstraints = false }

        itemView.fitIn(view: swipeView)

        if let superview = menuDotsView.superview {
            let menuDotsViewEdges = [
                superview.leadingAnchor.constraint(equalTo: menuDotsView.leadingAnchor),
                superview.topAnchor.constraint(equalTo: menuDotsView.topAnchor),
                superview.trailingAnchor.constraint(equalTo: menuDotsView.trailingAnchor),
                superview.bottomAnchor.constraint(equalTo: menuDotsView.bottomAnchor),
            ]

            NSLayoutConstraint.activate(menuDotsViewEdges)
        }
    }

    // MARK: - DrawerOverrides

    override func drawerScrollingStarts() {
        overscrollStartDate = nil
    }

    override func setVisualDrawerOffset(_ visualDrawerOffset: CGFloat, updateUI doUpdate: Bool) {
        super.setVisualDrawerOffset(visualDrawerOffset, updateUI: doUpdate)

        // After X % of reveal we consider animation should be finished
        let progress = visualDrawerOffset / SwipeMenuCollectionCell.MaxVisualDrawerOffsetRevealDistance
        menuDotsView.setProgress(progress, animated: true)
        if progress >= 1, overscrollStartDate == nil {
            overscrollStartDate = Date()
        }

        itemView.visualDrawerOffset = visualDrawerOffset
    }

    func updateAppearance() {
        itemView.update(for: conversation)
    }

    func size(inCollectionViewSize collectionViewSize: CGSize) -> CGSize {
        if !ConversationListCell.cachedSize.equalTo(CGSize.zero),
           ConversationListCell.cachedSize.width == collectionViewSize.width {
            return ConversationListCell.cachedSize
        }

        let fullHeightString = "Ü"
        itemView.configure(
            with: NSAttributedString(string: fullHeightString),
            subtitle: NSAttributedString(string: fullHeightString, attributes: ZMConversation.statusRegularStyle())
        )

        let fittingSize = CGSize(width: collectionViewSize.width, height: 0)

        itemView.frame = CGRect(x: 0, y: 0, width: fittingSize.width, height: 0)

        var cellSize = itemView.systemLayoutSizeFitting(fittingSize)
        cellSize.width = collectionViewSize.width
        ConversationListCell.cachedSize = cellSize
        return cellSize
    }

    static func invalidateCachedCellSize() {
        cachedSize = CGSize.zero
    }

    @objc
    private func onRightAccessorySelected(_: UIButton?) {
        guard let conversation = conversation as? ZMConversation else { return }

        let activeMediaPlayer = AppDelegate.shared.mediaPlaybackManager?.activeMediaPlayer

        if activeMediaPlayer != nil,
           activeMediaPlayer?.sourceMessage?.conversationLike === conversation {
            toggleMediaPlayer()
        } else if conversation.canJoinCall {
            delegate?.conversationListCellJoinCallButtonTapped(self)
        }
    }

    func toggleMediaPlayer() {
        let mediaPlaybackManager = AppDelegate.shared.mediaPlaybackManager

        if mediaPlaybackManager?.activeMediaPlayer?.state == .playing {
            mediaPlaybackManager?.pause()
        } else {
            mediaPlaybackManager?.play()
        }

        updateAppearance()
    }

    // MARK: - ConversationChangeInfo

    func setupConversationObserver(conversation: ZMConversation) {
        conversationObserverToken = ConversationChangeInfo.add(observer: self, for: conversation)
    }
}

// MARK: ZMTypingChangeObserver

extension ConversationListCell: ZMTypingChangeObserver {
    func typingDidChange(conversation: ZMConversation, typingUsers: [UserType]) {
        updateAppearance()
    }
}

// MARK: AVSMediaManagerClientObserver

extension ConversationListCell: AVSMediaManagerClientObserver {
    func mediaManagerDidChange(_ notification: AVSMediaManagerClientChangeNotification?) {
        guard !ProcessInfo.processInfo.isRunningTests else { return }

        // AUDIO-548 AVMediaManager notifications arrive on a background thread.
        DispatchQueue.main.async {
            if notification?.microphoneMuteChanged != nil {
                self.updateAppearance()
            }
        }
    }
}
